src/permissions/permissions.service.ts

```diff
@@ -126,4 +146,192 @@ export class PermissionsService {
     }
     return false;
   }
+
+  /**
+   * @async
+   * Update a notes permissions.
+   * @param {Note} note - the note
+   * @param {NotePermissionsUpdateDto} newPermissions - the permissions the not should be set to
+   * @return {Note} the note with the new permissions
+   * @throws {NotInDBError} there is no note with this id or alias
+   * @throws {PermissionsUpdateInconsistentError} the new permissions specify a user or group twice.
+   */
+  async updateNotePermissions(
+    note: Note,
+    newPermissions: NotePermissionsUpdateDto,
+  ): Promise<Note> {
+    const users = newPermissions.sharedToUsers.map(
+      (userPermission) => userPermission.username,
+    );
+
+    const groups = newPermissions.sharedToGroups.map(
+      (groupPermission) => groupPermission.groupName,
+    );
+
+    if (checkArrayForDuplicates(users) || checkArrayForDuplicates(groups)) {
+      this.logger.debug(
+        `The PermissionUpdate requested specifies the same user or group multiple times.`,
+        'updateNotePermissions',
+      );
+      throw new PermissionsUpdateInconsistentError(
+        'The PermissionUpdate requested specifies the same user or group multiple times.',
+      );
+    }
+
+    note.userPermissions = Promise.resolve([]);
+    note.groupPermissions = Promise.resolve([]);
+
+    // Create new userPermissions
+    for (const newUserPermission of newPermissions.sharedToUsers) {
+      const user = await this.usersService.getUserByUsername(
+        newUserPermission.username,
+      );
+      const createdPermission = NoteUserPermission.create(
+        user,
+        note,
+        newUserPermission.canEdit,
+      );
+      createdPermission.note = note;
+      (await note.userPermissions).push(createdPermission);
+    }
+
+    // Create groupPermissions
+    for (const newGroupPermission of newPermissions.sharedToGroups) {
+      const group = await this.groupsService.getGroupByName(
+        newGroupPermission.groupName,
+      );
+      const createdPermission = NoteGroupPermission.create(
+        group,
+        note,
+        newGroupPermission.canEdit,
+      );
+      createdPermission.note = note;
+      (await note.groupPermissions).push(createdPermission);
+    }
+
+    return await this.noteRepository.save(note);
+  }
+
+  /**
+   * @async
+   * Set permission for a specific user on a note.
+   * @param {Note} note - the note
+   * @param {User} permissionUser - the user for which the permission should be set
+   * @param {boolean} canEdit - specifies if the user can edit the note
+   * @return {Note} the note with the new permission
+   */
+  async setUserPermission(
+    note: Note,
+    permissionUser: User,
+    canEdit: boolean,
+  ): Promise<Note> {
+    const permissions = await note.userPermissions;
+    const permission = permissions.find(
+      (value: NoteUserPermission, index: number) => {
+        if (value.user.id == permissionUser.id) {
+          if (value.canEdit != canEdit) {
```

Comment:
Changing the value *inside* the `find` is quite confusing.
Please move that below the predicate.

Reply:
fixed